### PR TITLE
Improvement: Add support for target elementId in callout/contextualmenu

### PR DESCRIPTION
**Pull request by joschect**

### CODE REVIEW

**Comment by joschect** [comment]
It looks like the props changed a lot but I was streamlining them, they no longer rely on the positionprops.

File: src/components/Callout/Callout.Props.ts [hunk 1]
```diff
@@ -2,45 +2,71 @@ import * as React from 'react';
...
+   * How the element should be positioned
+   * @default DirectionalHint.BottomAutoEdge
    */
   directionalHint?: DirectionalHint;
 
   /**
-   * The gap space between the target element and the callout.
-   * @default 16
+   * The gap between the Callout and the target
+   * @default 0
```
Comment (Jahnp) [inline src/components/Callout/Callout.Props.ts]: Why is this being reverted? The original design spec called for 16 as default.

Comment (joschect) [inline src/components/Callout/Callout.Props.ts]: Whoops! That was a mistake, let me change it back.

Comment (Jahnp) [inline src/components/Callout/Callout.Props.ts]: Cool, thanks!

File: src/components/ContextualMenu/ContextualMenu.Props.ts [hunk 2]
```diff
@@ -1,35 +1,73 @@
...
+   * The ContextualMenu will then use document.getElementById to get the element.
    */
-  targetElement?: HTMLElement;
+  targetElement?: HTMLElement | string;
 
   /**
-   * Indicator of how the ContextualMenu should be anchored to its targetElement.
-   * @default DirectionalHint.rightBottomEdge
+   * How the element should be positioned
+   * @default DirectionalHint.BottomAutoEdge
```
Comment (dzearing) [inline src/components/ContextualMenu/ContextualMenu.Props.ts]: this is correct but ContextulaMenu.defaultProps.directionalHint is not.

File: src/components/Callout/CalloutContent.tsx [hunk 3]
```diff
@@ -52,9 +53,16 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
       slideDirectionalClassName: null,
       calloutElementRect: null
     };
+
+    if (typeof props.targetElement === 'string') {
+      this._targetElement = document.getElementById(props.targetElement);
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: Hmmm... this is a bad idea on multiple levels.

1. You're caching the result in c'tor. If the targetElement value ever changes it never recomputes.
2. You're referring to "document" in ctor, which is violating the SSR requirement of not assuming you have a browser. I'm surprised this didn't break the SSR test...
3. You're using document... I recall there was a helper for getting document to support multiple document scenarios.

File: src/components/ContextualMenu/ContextualMenu.Props.ts (hunk 2 above)
Comment (joschect) [inline src/components/ContextualMenu/ContextualMenu.Props.ts]: Okay got that fixed!

File: src/components/Callout/CalloutContent.tsx (hunk 3 above)
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: ping, just a reminder. I'd love to get this in. :)

File: src/components/Callout/Callout.Props.ts [hunk 4]
```diff
@@ -2,45 +2,71 @@ import * as React from 'react';
...
-   * @default true
+   * If true use a point rather than rectangle to position the Callout.
+   * For example it can be used to position based on a click.
+   */
+  useTargetPoint?: boolean;
+
+  /**
+   * Point used to position the Callout
+   */
+  targetPoint?: IPoint;
```
Comment (dzearing) [inline src/components/Callout/Callout.Props.ts]: this is rather confusing... relative to... closest layer host?

File: src/components/Callout/CalloutContent.tsx [hunk 5]
```diff
@@ -21,212 +23,240 @@ const OFF_SCREEN_POSITION = { top: -9999, left: 0 };
...
-      this._targetWindow = window;
+
+    private _didSetInitialFocus: boolean;
+    private _hostElement: HTMLDivElement;
+    private _calloutElement: HTMLDivElement;
+    private _targetWindow: Window;
+    private _bounds: IRectangle;
+    private _maxHeight: number;
+    private _positionAttempts: number;
+    private _target: HTMLElement | MouseEvent;
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: or string?

File: src/components/Callout/Callout.Props.ts [hunk 6]
```diff
@@ -2,45 +2,75 @@ import * as React from 'react';
 import { Callout } from './Callout';
 import { CalloutContent } from './CalloutContent';
 import { DirectionalHint } from '../../common/DirectionalHint';
-import { IPositionProps } from '../../utilities/positioning';
+import { IRectangle } from '../../common/IRectangle';
+import { IPoint } from '../../common/IPoint';
+
+export interface ICalloutProps extends React.Props<Callout | CalloutContent> {
+
+  target?: HTMLElement | string | MouseEvent;
```
Comment (dzearing) [inline src/components/Callout/Callout.Props.ts]: no comments

**Comment by dzearing** [comment]
Closing this per discussion until we get this right

**Pull request closed without merging**